sympy/physics/units/util.py

```diff
@@ -25,60 +21,54 @@ def dim_simplify(expr):
     dimension without being a dimension. This is necessary to avoid strange
     behavior when Add(L, L) be transformed into Mul(2, L).
     """
-
-    if isinstance(expr, Dimension):
-        return expr
-
-    if isinstance(expr, Pow):
-        return dim_simplify(expr.base)**dim_simplify(expr.exp)
-    elif isinstance(expr, Function):
-        return dim_simplify(expr.args[0])
-    elif isinstance(expr, Add):
-        if (all(isinstance(arg, Dimension) for arg in expr.args) or
-            all(arg.is_dimensionless for arg in expr.args if isinstance(arg, Dimension))):
-            return reduce(lambda x, y: x.add(y), expr.args)
-        else:
-            raise ValueError("Dimensions cannot be added: %s" % expr)
-    elif isinstance(expr, Mul):
-        return Dimension(Mul(*[dim_simplify(i).name for i in expr.args if isinstance(i, Dimension)]))
-
-    raise ValueError("Cannot be simplifed: %s", expr)
+    _, expr = Quantity._collect_factor_and_dimension(expr)
+    return expr
 
 
 def _get_conversion_matrix_for_expr(expr, target_units):
     from sympy import Matrix
 
     expr_dim = Dimension(Quantity.get_dimensional_expr(expr))
-    dim_dependencies = expr_dim.get_dimensional_dependencies(mark_dimensionless=True)
-    target_dims = [Dimension(Quantity.get_dimensional_expr(x)) for x in target_units]
-    canon_dim_units = {i for x in target_dims for i in x.get_dimensional_dependencies(mark_dimensionless=True)}
+    dim_dependencies = expr_dim.get_dimensional_dependencies(
+        mark_dimensionless=True)
+    target_dims = [
+        Dimension(Quantity.get_dimensional_expr(x)) for x in target_units
+    ]
+    canon_dim_units = {
+        i
+        for x in target_dims
+        for i in x.get_dimensional_dependencies(mark_dimensionless=True)
+    }
```

Comment:
There is a readability issue if you don't break the line since `i` is not used on `for x in target_dims`.